.mage/proto.go

```diff
@@ -0,0 +1,224 @@
+		if err != nil {
+			return err
+		}
+		for _, ext := range []string{".pb.go", ".pb.gw.go", ".pb.fm.go", ".pb.util.go"} {
+			if strings.HasSuffix(path, ext) {
+				if err := sh.Rm(path); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+		return nil
+	})
+}
+
+// Swagger generates Swagger protos.
+func (p Proto) Swagger(context.Context) error {
+	changed, err := target.Glob(filepath.Join("api", "api.swagger.json"), filepath.Join("api", "*.proto"))
+	if err != nil {
+		return errfmt.Errorf("failed checking modtime: %w", err)
+	}
+	if !changed {
+		return nil
+	}
+	return withProtoc(func(pCtx *protocContext, protoc func(...string) error) error {
+		if err := protoc(
+			fmt.Sprintf("--swagger_out=allow_merge,merge_file_name=api:%s/api", pCtx.WorkingDirectory),
+			fmt.Sprintf("%s/api/*.proto", pCtx.WorkingDirectory),
+		); err != nil {
+			return errfmt.Errorf("failed to generate protos: %w", err)
+		}
+		return nil
+	})
+}
+
+// SwaggerClean removes generated Swagger protos.
+func (p Proto) SwaggerClean(context.Context) error {
+	return sh.Rm(filepath.Join("api", "api.swagger.json"))
+}
+
+// Markdown generates Markdown protos.
+func (p Proto) Markdown(context.Context) error {
+	changed, err := target.Glob(filepath.Join("api", "api.md"), filepath.Join("api", "*.proto"))
+	if err != nil {
+		return errfmt.Errorf("failed checking modtime: %w", err)
+	}
+	if !changed {
+		return nil
+	}
+	return withProtoc(func(pCtx *protocContext, protoc func(...string) error) error {
+		if err := protoc(
+			fmt.Sprintf("--doc_opt=%s/api/api.md.tmpl,api.md --doc_out=%s/api", pCtx.WorkingDirectory, pCtx.WorkingDirectory),
+			fmt.Sprintf("%s/api/*.proto", pCtx.WorkingDirectory),
+		); err != nil {
+			return errfmt.Errorf("failed to generate protos: %w", err)
+		}
+		return nil
+	})
+}
+
+// MarkdownClean removes generated Markdown protos.
+func (p Proto) MarkdownClean(context.Context) error {
+	return sh.Rm(filepath.Join("api", "api.md"))
+}
+
+// JsSDK generates javascript SDK protos.
+func (p Proto) JsSDK(context.Context) error {
+	changed, err := target.Glob(filepath.Join("sdk", "js", "generated", "api.json"), filepath.Join("api", "*.proto"))
+	if err != nil {
+		return errfmt.Errorf("failed checking modtime: %w", err)
+	}
+	if !changed {
+		return nil
+	}
+	return withProtoc(func(pCtx *protocContext, protoc func(...string) error) error {
+		if err := protoc(
+			fmt.Sprintf("--doc_opt=json,api.json --doc_out=%s/sdk/js/generated", pCtx.WorkingDirectory),
+			fmt.Sprintf("%s/api/*.proto", pCtx.WorkingDirectory),
+		); err != nil {
+			return errfmt.Errorf("failed to generate protos: %w", err)
+		}
+		return nil
+	})
+}
+
+// JsSDKClean removes generated javascript SDK protos.
+func (p Proto) JsSDKClean(context.Context) error {
+	return sh.Rm(filepath.Join("sdk", "js", "generated", "api.json"))
+}
+
+// All generates protos.
+func (p Proto) All(ctx context.Context) {
+	mg.CtxDeps(ctx, Proto.Image)
```

Comment:
Remove this.
1. It's redundant, since `docker run` will automatically pull the image if not available.
2. It breaks `proto:all` in offline scenarios in cases where the image is already available and in cases where image is available locally, but not in the repository.
3. This target generates protos. Updating the image is an unexpected side-effect.

Reply:
When I ran this `proto:all` without pulling the image first, it started pulling the image concurrently for all the jobs below. We don't want that. Feel free to suggest an alternative.

Reply:
I favor offline support over concurrent pulling images. If we can have (explicit) offline support, I'm fine with that too, but we need offline support.

Reply:
Fixed that in https://github.com/TheThingsNetwork/lorawan-stack/pull/377/commits/82c670f81398aaea1d8b7b740543353ab0c7922b